iml-agent/src/daemon_plugins/ntp.rs

```diff
@@ -0,0 +1,340 @@
+
+        match ntp_synced {
+            Some(time::Synced::Synced) => time::State::Synced,
+            Some(time::Synced::Unsynced) => time::State::Unsynced(ntp_offset),
+            None => time::State::Unknown,
+        }
+    } else {
+        let (chrony_synced, chrony_offset) = chronyd_synced().await?;
+
+        match chrony_synced {
+            Some(time::Synced::Synced) => time::State::Synced,
+            Some(time::Synced::Unsynced) => time::State::Unsynced(chrony_offset),
+            None => time::State::Unknown,
+        }
+    };
+
+    let x = serde_json::to_value(state).map(Some)?;
+
+    Ok(x)
+}
+
+impl DaemonPlugin for Ntp {
+    fn start_session(
+        &mut self,
+    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
+        self.update_session()
+    }
+
+    fn update_session(
+        &self,
+    ) -> Pin<Box<dyn Future<Output = Result<Output, ImlAgentError>> + Send>> {
+        Box::pin(time_synced(
+            is_ntp_configured_by_iml,
+            get_time_sync_services,
+            ntpd_synced,
+            chronyd_synced,
+        ))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use futures::FutureExt;
+
+    #[test]
+    fn test_is_ntp_synced() {
+        let s = r#"b true"#;
+
+        assert_eq!(parse_ntp_synced(s), Some(time::Synced::Synced));
+    }
+
+    #[test]
+    fn test_is_ntp_not_synced() {
+        let s = r#"b false"#;
+
+        assert_eq!(parse_ntp_synced(s), Some(time::Synced::Unsynced));
+    }
+
+    #[test]
+    fn test_bad_output() {
+        let s = r#"b tr"#;
+
+        assert_eq!(parse_ntp_synced(s), None);
+    }
+
+    #[test]
+    fn test_get_ntp_time_offset() {
+        let s = r#"synchronised to NTP server (10.73.10.10) at stratum 12
+time correct to within 949 ms
+polling server every 64 s"#;
+
+        assert_eq!(parse_ntp_time_offset(s), Some("949 ms".to_string().into()));
+    }
+
+    #[test]
+    fn test_get_chrony_time_offset() {
+        let s = r#"
+
+Remote address  : 10.73.10.10 (0A490A0A)
+Remote port     : 123
+Local address   : 10.73.10.12 (0A490A0C)
+Leap status     : Normal
+Version         : 4
+Mode            : Server
+Stratum         : 11
+Poll interval   : 4 (16 seconds)
+Precision       : -25 (0.000000030 seconds)
+Root delay      : 0.000000 seconds
+Root dispersion : 0.011124 seconds
+Reference ID    : 7F7F0100 ()
+Reference time  : Fri Sep 13 08:03:04 2019
+Offset          : +0.000026767 seconds
+Peer delay      : 0.000403893 seconds
+Peer dispersion : 0.000000056 seconds
+Response time   : 0.000079196 seconds
+Jitter asymmetry: +0.00
+NTP tests       : 111 111 1111
+Interleaved     : No
+Authenticated   : No
+TX timestamping : Kernel
+RX timestamping : Kernel
+Total TX        : 1129
+Total RX        : 1129
+Total valid RX  : 1129"#;
+
+        assert_eq!(
+            parse_chrony_time_offset(s),
+            Some("0.000026767 seconds".to_string().into())
```

Comment:
Likewise here. Now it looks like you are using the same parser on both sides. And actually you are, you only use a regex to extract a string, and then parse the string the same way on both sides. Tautological :)